coreos-iso-embed-ignition

```diff
@@ -0,0 +1,178 @@
+            FILENAME], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+    config, _ = proc.communicate(cpio)
+    if proc.returncode != 0:
+        raise EmbedException(f'cpio -i exited with status {proc.returncode}.')
+    return config
+
+
+def get_location(iso_fh):
+    '''
+    Given an ISO image, return the offset and length of the Ignition
+    embed area.
+    '''
+    fmt = '<8s2Q'
+    size = struct.calcsize(fmt)
+    iso_fh.seek(32768 - size)
+    buf = iso_fh.read(size)
+    if len(buf) < 24:
+        raise EmbedException('Short ISO image file.')
+    magic, offset, length = struct.unpack(fmt, buf)
+    if magic != b'coreiso+':
+        raise EmbedException('Unrecognized CoreOS ISO image.')
+    if os.fstat(iso_fh.fileno()).st_size < offset + length:
+        raise EmbedException('Invalid CoreOS ISO image.')
+    return offset, length
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Embed an Ignition config in a CoreOS live ISO image.")
+    subparsers = parser.add_subparsers(dest='command', title='subcommands',
+            required=True,
+            description='Use <subcommand> -h for more info.')
+
+    embed = subparsers.add_parser('embed',
+            description='Embed an Ignition config in an ISO image.',
+            help='Embed an Ignition config in an ISO image')
+    embed.add_argument('-c', '--config', metavar='PATH', default='-',
+            type=argparse.FileType('r'),
+            help='Ignition config to embed [default: stdin]')
+    embed.add_argument('-f', '--force', action='store_true',
+            help='Overwrite an existing embedded Ignition config')
+    embed.add_argument('-o', '--output', metavar='PATH',
+            type=argparse.FileType('x+b'),
+            help='Copy to a new file, instead of modifying in place')
+    embed.set_defaults(command=do_embed)
+
+    show = subparsers.add_parser('show',
+            description='Show the embedded Ignition config from an ISO image.',
+            help='Show the embedded Ignition config from an ISO image')
+    show.set_defaults(command=do_show)
+
+    remove = subparsers.add_parser('remove',
+            description='Remove an existing embedded Ignition config from an ISO image.',
+            help='Remove an existing embedded Ignition config')
+    remove.add_argument('-o', '--output', metavar='PATH',
+            type=argparse.FileType('x+b'),
+            help='Copy to a new file, instead of modifying in place')
+    remove.set_defaults(command=do_remove)
+
+    for subp in embed, show, remove:
+        subp.add_argument('image', metavar='ISO',
+                type=argparse.FileType('r+b'),
+                help='ISO image')
+
+    return parser.parse_args()
+
+
+def do_embed(args, fh, offset, length):
+    cpio = make_cpio(args.config)
+    if len(cpio) > length:
+        raise EmbedException('Compressed Ignition config is too large: {} > {}'.format(len(cpio), length))
```

Comment:
how large is too large? we should probably document and also tell the user here

Reply:
We tell the user what we can; see the last part of the message.  We don't know how well their Ignition config will compress, though.